Fix buffer size issue for AF_NETLINK

Leseratte10 opened the pull request:
This should hopefully fix #5 in all cases, by almost completely rewriting getsockname and getpeername It does fix the issue with AF_NETLINK I had with my application in particular. 

It also fixes another issue with the buffer size. Previously, if the `__addr` buffer in `getsockname` or `getpeername` was too small for the struct, your code just aborted and returned -1. 

According to [the documentation](https://man7.org/linux/man-pages/man2/getsockname.2.html) though, the returned data is supposed to be truncated to fit into the buffer, and the real needed buffer size is supposed to be returned through `__len` so the calling application can either decide to retry with a larger buffer, or be satisfied with the partial response. 

Also, `getsockname` is the IP address the local side of the socket is working on/with (your IP). So if you open a socket and connect to a remote server, getsockname is supposed to return your own IP. What happens currently is that in this case the library just "throws" the IPv6 address back to the application, which is probably not going to end well when an IPv4-only application actually tries to use this. I've modded it to return the IPv4 address 0.0.0.0 instead for the local socket.

Leseratte10 committed:
Fix buffer size issue for AF_NETLINK

andrewshadura commented on tnat64.c:
```diff
@@ -415,29 +429,26 @@ int getsockname(GETSOCKNAME_SIGNATURE)
     /* If we haven't initialized yet, do it now */
     get_config();
 
+    /* The software calling getsockname is expecting an IPv4 response. 
+    This means that the __addr pointer might only have space for a sockaddr_in, 
+    not for a sockaddr_in6. It's probably unreasonable to expect the calling
+    application to provide a buffer large enough for an IPv6 sockaddr_in6
+    if they're assuming they talk IPv4. So, better allocate our own buffers, temporarily. */
+
```
Very good point.

andrewshadura commented on tnat64.c:
```diff
@@ -389,21 +388,36 @@ int getpeername(GETPEERNAME_SIGNATURE)
             return ret;
         }
+        else {
+            // Not sure what the best data to return here would be. 
+            // This should never happen in normal operation, though. 
+            // This would only be executed if the socket was connected to an IPv6 address 
+            // that's not part of a NAT64 prefix.
```
I guess, at least some
`show_msg(MSGWARN, "<...>");`

andrewshadura replied:
OTOH probably not, since even if this happens, anything should transparently work as before.

Leseratte10 replied:
I mean, yeah, everything should continue like with older versions of tnat64, but it's still an indicator that something's amiss. But same as with your other comment, I guess this can happen when an application opens IPv6 sockets natively?

Leseratte10 replied:
Added a log warning to the code if that happens.

andrewshadura commented on tnat64.c:
```diff
@@ -415,54 +429,91 @@ int getsockname(GETSOCKNAME_SIGNATURE)
+            // The socket is listening on a "real" IPv6 address. 
+            // Returning that address to the application as-is is going to cause issues. 
+            // It's probably better to make the application believe it is bound to the unspecific address, 
+            // i.e. return an IPv4 address of 0.0.0.0 in this case. 
```
Won’t this break applications talking to both IPv4 and IPv6?

Leseratte10 replied:
That's a very good point, that's probably true. I've also found another oddity in the code that could break applications that talk both IPv4 and IPv6. 

I guess the only way to 100% prevent these kinds of errors would be to keep a list of all sockets the application opened through the `socket` function, and take a note whether the socket originally was an IPv6 socket, or if it was an IPv4 socket that got "upgraded" to IPv6 through the library. 

Then it could, for "real" IPv6 sockets just return the original getsockname/getpeername, and for "fake" IPv6 sockets (where the application thinks it opened an IPv4 socket) it can fake the address 0.0.0.0. 
Does this sound like a sensible solution? Might also make it easier to add UDP support in the future where you'd probably have the same issue with sendto and recvfrom.

I guess that would also make it easier to handle issues like this one, with AF_NETLINK. You could just, on every single function call, first check if the fd belongs to A) a native IPv6 socket (straight through to the original function), B) a fake IPv6 socket (where we need to spoof an IPv4, or C) some other protocol socket like AF_NETLINK (straight through to the original function, too).

Leseratte10 replied:
This issue should now be fixed with the updated patch, see my comment below.

Leseratte10 committed:
Improve compatibility with real IPv6 sockets

Leseratte10 commented:
I've updated the code a bit with the following things: 

- When execution starts, I'm allocating an array named `socket_fd_flags`. That array has four bits of space for each file descriptor. Then, when the application creates an IPv4 socket and tnat64 "upgrades" that to an IPv6 socket, the flag `TNAT_FLAG_SOCK_UPGRADED_TO_IPV6` is set. If the application already requested an IPv6 socket, that flag is not set. That way, all the code will know if the socket was intended to be a native IPv6 socket or if the application thinks it's an IPv4 socket. The functions `get_custom_fd_flags` and `set_custom_fd_flags` can be used to access that array.
(The three other bits are empty and could be used in the future when you need to store more information for each socket).
- Currently, the code assumes that it can always send IPv4 packets through an IPv6 socket - but some systems may not support dual-stack sockets, or an admin deliberately disabled them. With this change, the code calling "is_local" handling local IPv4 connections will only do that if the IPv6 socket supports sending packets to IPv4 destinations; same for the code for the NAT64-to-IPv4-fallback (at the end of `connect`).
- If a connection to 127.0.0.1 fails because the socket only accepts connections to IPv6 destinations, it now tries to connect to ::1 instead.
- `getpeername` and `getsockname` check the socket flags mentioned before, and if the flag `TNAT_FLAG_SOCK_UPGRADED_TO_IPV6` is *NOT* set, it immediately passes execution through to the original implementation of `getpeername` or `getsockname` so the rest of the code, including the IP address shenanigans, only occur for IPv4 sockets that the library upgraded to IPv6. That also makes it less likely that bugs in the library break rarely-tested types like AF_NETLINK or AF_UNIX again.

andrewshadura commented:
Wow, this is huge 🙂

I’m thinking, if we’re going this big, wouldn’t it better to rewrite it in Rust? Doing what you’re trying to do in this PR would be much easier. I tried prototyping something last weekend, and I’ve got some things working, but I’m super early in it, I only wrap the four functions and add debug prints.

How’s your Rust? Would you be able to contribute to a Rust codebase?

Leseratte10 commented:
How's my Rust? Well, I know how to spell the word "Rust" and I know that it's a programming language, that's about it ^^ Not sure how easy it would be to get into Rust to a level to be able to make helpful PRs. Although, if you have something working in Rust, couldn't hurt to have both a C and a Rust variant of the library. I do know that any other programming language other than C is probably way easier to write safe code that handles all corner cases.

I was planning, once this PR is merged, to continue with #6 (different prefix sizes), and once that's done, try adding support to autodiscover the NAT64 prefix (RFC7050). And then, if possible, UDP support, and then the library should be fairly feature-complete.

Leseratte10 commented:
So, any status update regarding this? From my side, this PR should be done unless you have further comments regarding something I should change. Would you be willing to merge this so I can continue working on the other stuff I've planned to implement (prefix autodiscover, UDP, ...); or do you plan to no longer maintain this version in favour of a Rust rewrite? If so, is that already available somewhere for me to take a look at?

andrewshadura commented:
Hi,
Sorry, I haven’t had time for this yet. The Rust rewrite was just a prototype and I don’t expect it to be ready soon yet.
I’ll try to have a look over the weekend.

Leseratte10 commented:
Thanks for the response. No problem, I just wasn't sure if you were waiting for another response from me, that's why I asked.

Leseratte10 committed:
Add support to resize the flags array if the fd limit changes

Leseratte10 commented:
I've added two more changes: 

- The socket flag array is now allocated upon first usage, rather than at library start. This means that if the library is loaded into a program that doesn't do networking, or a program that's fully IPv6-aware, it won't waste RAM space for that never-used table. 
- Applications running as root are allowed to increase / decrease the file descriptor limit, so just allocating once at program start isn't enough. Now the array is re-allocated if it's too small because the limit has increased. 

I hope all that new malloc/realloc code is correct ...

Leseratte10 commented:
@andrewshadura Any update yet?

Leseratte10 commented:
Is there any update yet @andrewshadura ? Do you plan to review and merge this eventually, or would I be better off maintaining my own fork of tnat64?

andrewshadura commented:
I’m really sorry for delaying it for this long. Adding this to my calendar to look at tomorrow.

andrewshadura commented on tnat64.c:
```diff
@@ -324,8 +506,29 @@ int connect(CONNECT_SIGNATURE)
+                    int sockopt = -1;
+                    socklen_t len = sizeof(sockopt);
+                    if (getsockopt(__fd, IPPROTO_IPV6, IPV6_V6ONLY, (void*)&sockopt, &len) < 0) {
+                        show_msg(MSGWARN, "Can't figure out if this IPv6 socket supports IPv4, assume yes - error %d (%s)\n", errno, strerror(errno));
```
```suggestion
                        show_msg(MSGWARN, "Can't figure out if this IPv6 socket supports IPv4, assuming it does (error %d: %s)\n", errno, strerror(errno));
```

Leseratte10 committed:
Fixups after review

Co-authored-by: Andrej Shadura <andrew@shadura.me>

andrewshadura commented:
So, there’s a lot of code that seemingly looks okay, but I didn’t review in depth yet, but I’m thinking — we need to have a way to test it properly. So I’ve spent some hours yesterday adding a simple test based on #1 and #3, and also tried out Meson as a build system 🙂

Leseratte10 commented:
Looks like one of your tests is already failing in this branch. Question is, is my code broken or is your test broken? :)

Leseratte10 commented:
Looks like the test is broken. With my changes it just no longer hits that code path. What was that test (`01-correct-ipv6-address`) supposed to test? A test for the invalid IPv6s I saw in #1 ?

andrewshadura commented:
Well, yeah, it was not supposed to hit this bug, you need to add a new test failing on the current main but passing in this branch 🙂

-- 
Cheers,
  Andrej

Leseratte10 commented:
Ah, so basically just revert the existing test? Fail if it finds that string, success otherwise?

Leseratte10 committed:
Update test

andrewshadura commented:
Wait, what bugs are we talking about? This test was just for the garbage in the logs, but not for the netlink issue.

Leseratte10 commented:
When executing this test, the current `main` version prints this: 

```
16:48:39 libtnat64(1610858): Opening configuration file (/etc/tnat64.conf)
16:48:39 libtnat64(1610858): New network entry for 10.0.0.0
16:48:39 libtnat64(1610858): New network entry for 127.0.0.0
16:48:39 libtnat64(1610858): Got getsockname call for socket 3
16:48:39 libtnat64(1610858): Address family is AF_INET6
16:48:39 libtnat64(1610858): Checking if IPv6 address :: is behind the NAT64...
16:48:39 libtnat64(1610858): Checking the default NAT64 prefix 64:ff9b::
16:48:39 libtnat64(1610858): Got getsockname call for socket 6
16:48:39 libtnat64(1610858): Address family is AF_INET
16:48:39 libtnat64(1610858): Got getsockname call for socket 6
16:48:39 libtnat64(1610858): Address family is AF_INET6
16:48:39 libtnat64(1610858): Checking if IPv6 address ::1 is behind the NAT64...
16:48:39 libtnat64(1610858): Checking the default NAT64 prefix 64:ff9b::
```

(and the original test on main passes).

But the version inside this PR (#8) prints this: 

```
16:49:49 libtnat64(1611368): Checking file descriptor limits - current limit: 1024, max limit: 1048576
16:49:49 libtnat64(1611368): Perform initial flag array allocation (size 524289 bytes, 1048576 entries)
16:49:49 libtnat64(1611368): Allocated the flag buffer to size of 524289 (1048576 entries)!
16:49:49 libtnat64(1611368): Opening configuration file (/etc/tnat64.conf)
16:49:49 libtnat64(1611368): New network entry for 10.0.0.0
16:49:49 libtnat64(1611368): New network entry for 127.0.0.0
16:49:49 libtnat64(1611368): Got getsockname call for socket 3
16:49:49 libtnat64(1611368): None of our modded sockets, call real getsockname
16:49:49 libtnat64(1611368): Got getsockname call for socket 6
16:49:49 libtnat64(1611368): None of our modded sockets, call real getsockname
16:49:49 libtnat64(1611368): Got getsockname call for socket 6
16:49:49 libtnat64(1611368): None of our modded sockets, call real getsockname
```
(so the test fails)

Because "dig" is opening a UDP socket for DNS queries, and tnat64 is (currently) incapable of nat64'ing UDP connections, the calls skip most of the code and don't even run into the code path that had the issue with the corrupted logging which is tested by that test. So the test with dig is essentially useless. For a real test you'd need to use TCP, but I don't think an IP of 0.0.0.0 / :: will work there either.

EDIT: I do know that I should still add a whole new test for the NETLINK issue in particular, but that's a different topic. This change / question was about the existing test (for bug #1) failing on this branch because the code that's being tested here is no longer even being executed.

andrewshadura commented:
Ah, okay, I see.

Leseratte10 committed:
Add a test to ensure NETLINK works

Leseratte10 commented:
I just added a third test that verifies that netlink sockets can call getsockname without errors. This test fails on `main` but succeeds with this PR. 

I hope a dependency on python for that test isn't an issue; I made sure to test it with python2 and python3; but that was the easiest way I could figure out to test a netlink socket.

Leseratte10 committed:
Fix issue where nonstandard values for IPV6_V6ONLY would fail

IPV6_V6ONLY is only supposed to have the values 0 or 1.
However, the man page says to assume 0=false and everything else = true.

Leseratte10 commented:
Any update yet? Is there something I still need to do, or do you just need to review this? Or do you want to switch to the new meson build system?

I'm hoping to be able to add a couple more features as well (UDP, auto-discovering the NAT64 prefix, ...) but I'd rather not have too many open PRs with different features because I don't want to run into more merge conflicts.
